src/QVContainer.tsx

```diff
@@ -1,87 +1,272 @@
+      const stagesCopy = [...this.state.Model.stages];
+      stagesCopy[index].policy = policy;
+      modelCopy.stages = stagesCopy;
+      this.setState({ Model: modelCopy });
+    }
+
+    private handleForbidUndervoteChange(val: string, index: number): void {
+      const modelCopy = { ...this.state.Model };
+      const stagesCopy = [...this.state.Model.stages];
+      stagesCopy[index].forbidUndervote = val === "1";
+      modelCopy.stages = stagesCopy;
+      this.setState({ Model: modelCopy });
+    }
 
-  // Called by PluginMain() once sheetId is available.
-  public setSheetRef(sheetRef: IPluginSheetRef): void {
-    // debugger;
+    private handleNWinnersChange(num: string, index: number): void {
+      const modelCopy = { ...this.state.Model };
+      const stagesCopy = [...this.state.Model.stages];
+      if (stagesCopy[index].sourceInline === "Yes,No") {
+        alert(
+          'Invalid operation: the "Yes/No" source type supports at most one winner. Change the source type to increase the number of potential winners.'
+        );
+        return;
+      }
+      stagesCopy[index].nWinners = Number(num);
+      modelCopy.stages = stagesCopy;
+      this.setState({ Model: modelCopy });
+    }
+
+    private handleSInlineChange(val: string, index: number): void {
+      const modelCopy = { ...this.state.Model };
+      const stagesCopy = [...this.state.Model.stages];
+      stagesCopy[index].sourceInline = val;
+      modelCopy.stages = stagesCopy;
+      this.setState({ Model: modelCopy });
+    }
+
+    private handleSSlateChange(val: string, index: number): void {
+      const modelCopy = { ...this.state.Model };
+      const stagesCopy = [...this.state.Model.stages];
+      stagesCopy[index].sourceSlate = val;
+      modelCopy.stages = stagesCopy;
+      this.setState({ Model: modelCopy });
+    }
+
+    private handleSouceChange(val: string, index: number): void {
+      const modelCopy = { ...this.state.Model };
+      const stagesCopy = [...this.state.Model.stages];
+
+      if (val === "yn") {
+        stagesCopy[index].nWinners = 1;
+        stagesCopy[index].sourceInline = "Yes,No";
+        stagesCopy[index].sourceAlternates = false;
+        stagesCopy[index].sourceSlate = null;
+      }
+
+      if (val === "inline") {
+        stagesCopy[index].sourceInline = "";
+        stagesCopy[index].sourceAlternates = false;
+        stagesCopy[index].sourceSlate = null;
+      }
+
+      if (val === "slate") {
+        stagesCopy[index].sourceInline = null;
+        stagesCopy[index].sourceAlternates = false;
+        stagesCopy[index].sourceSlate = "";
+      }
+
+      if (val === "alternates") {
+        stagesCopy[index].sourceInline = null;
+        stagesCopy[index].sourceAlternates = true;
+        stagesCopy[index].sourceSlate = null;
+      }
+
+      modelCopy.stages = stagesCopy;
+      this.setState({ Model: modelCopy });
+    }
 
-    var server = sheetRef.Server;
-    // server = "https://localhost:44321";
```

Comment:
Can we leave this localhost line in even though it's commented out.  I know it's super hacky, but it helps me with debugging. ;)